src/wasm/wasm-validator.cpp

```diff
@@ -2658,7 +2658,7 @@ void FunctionValidator::visitTryTable(TryTable* curr) {
                 "the number of catch tags and sent types do not match");
 
   const char* invalidSentTypeMsg = "invalid catch sent type information";
-  Type exnref = Type(HeapType::exn, Nullable);
+  Type exnref = Type(HeapType::exn, NonNullable);
```

Comment:
I think we should be sending a non-nullable exnref. The target might be less refined (until ReFinalize fixes that) but we should still be sending the refined type, as the spec says?

Unless I missed something in the code here and it is checking something other than the type being sent?

Reply:
If I'm misunderstanding you, can you give an example of code that should validate but this code will not accept?

Reply:
I think I might have misunderstood what this PR does. So basically, from now on we assume `exnref` == `(ref exn)` and not `(ref null exn)` and treat it as such in the IR and all passes, and make it `(ref null exn)` only when we write it back to a binary. Right? (By the way don't we need to do the same thing when we write it back as a text?)

If that's the case, if you programatically create a `TryTable` object with `exnref` in `sentType` field, it will not validate. Of course, we don't usually manually set the fields in the IR classes; we usually use constructors and `finalize` methods, so I guess practically it's fine, but it was a little confusing to me. (Maybe giving `TryTable::sentType` a `private` visibility might help a little.)

I think it'd help if we have little more comments in wasm.cpp that why we are doing this. Also, can't we also validate `(ref null exn)` in `sentType` as valid? (I can't give you an example wast program that does not validate, because we basically read `exnref` as `(ref exn)` when reading.)

Reply:
`exnref` should be equivalent to `(ref null exn)`, not `(ref exn)`. Since `try_table` never produces a null exn value, the most precise type for the exn value it produces is always `(ref exn)`. We can therefore take it as an invariant that the IR models sending a `(ref exn)`. However, non-nullable references like that do not exist without GC enabled, so without GC, we fix it up by transforming the `(ref exn)` into `exnref` (i.e. `(ref null exn)`) in the binary writer.

Reply:
I added a comment now. I also noticed we were missing docs on this in the readme and added those.